File: app/src/main/java/dev/ragnarok/fenrir/activity/ValidateActivity.java
package dev.ragnarok.fenrir.activity;

import static dev.ragnarok.fenrir.util.Utils.nonEmpty;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.util.Log;
import android.webkit.CookieManager;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import dev.ragnarok.fenrir.AccountType;
import dev.ragnarok.fenrir.Constants;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.activity.slidr.Slidr;
import dev.ragnarok.fenrir.activity.slidr.model.SlidrConfig;
import dev.ragnarok.fenrir.api.Auth;
import dev.ragnarok.fenrir.api.util.VKStringUtils;
import dev.ragnarok.fenrir.model.Token;
import dev.ragnarok.fenrir.settings.CurrentTheme;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.settings.theme.ThemesController;
import dev.ragnarok.fenrir.util.Logger;

public class ValidateActivity extends AppCompatActivity {

    private static final String TAG = ValidateActivity.class.getSimpleName();
    private static final String EXTRA_VALIDATE = "validate";

    public static Intent createIntent(Context context, String validate_url) {
        return new Intent(context, LoginActivity.class)
                .putExtra(EXTRA_VALIDATE, validate_url);
    }

    private static String tryExtractAccessToken(String url) {
        return VKStringUtils.extractPattern(url, "access_token=(.*?)&");
    }

    private static ArrayList<Token> tryExtractAccessTokens(String url) throws Exception {
        Pattern p = Pattern.compile("access_token_(\\d*)=(.*?)&");

        ArrayList<Token> tokens = new ArrayList<>();

        Matcher matcher = p.matcher(url);
        while (matcher.find()) {
            String groupid = matcher.group(1);
            String token = matcher.group(2);

            if (nonEmpty(groupid) && nonEmpty(token)) {
                tokens.add(new Token(-Integer.parseInt(groupid), token));
            }
        }

        if (tokens.isEmpty()) {
            throw new Exception("Failed to parse redirect url " + url);
        }

        return tokens;
    }

    private static String tryExtractUserId(String url) {
        return VKStringUtils.extractPattern(url, "user_id=(\\d*)");
    }

    @SuppressLint("SetJavaScriptEnabled")
    @Override
    public void onCreate(Bundle savedInstanceState) {
        setTheme(ThemesController.INSTANCE.currentStyle());
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        Slidr.attach(this, new SlidrConfig.Builder().scrimColor(CurrentTheme.getColorBackground(this)).build());

        WebView webview = findViewById(R.id.vkontakteview);
        webview.getSettings().setJavaScriptEnabled(true);
        webview.clearCache(true);
        webview.getSettings().setUserAgentString(Constants.USER_AGENT(AccountType.KATE));

        //Чтобы получать уведомления об окончании загрузки страницы
        webview.setWebViewClient(new VkontakteWebViewClient());

        CookieManager cookieManager = CookieManager.getInstance();
        cookieManager.removeAllCookies(aBoolean -> Log.d(TAG, "Cookie removed: " + aBoolean));

        webview.loadUrl(getIntent().getStringExtra(EXTRA_VALIDATE));
    }

    private void parseUrl(String url) {
        try {
            if (url == null) {
                return;
            }

            Logger.d(TAG, "url=" + url);

            if (url.startsWith(Auth.redirect_url)) {
                if (!url.contains("error=")) {
                    Intent intent = new Intent();

                    try {
                        String accessToken = tryExtractAccessToken(url);
                        String userId = tryExtractUserId(url);
                        Settings.get().accounts().storeAccessToken(Integer.parseInt(userId), accessToken);
                    } catch (Exception ignored) {
                    }

                    setResult(RESULT_OK, intent);
                }

                finish();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private class VkontakteWebViewClient extends WebViewClient {
        @Override
        public void onPageStarted(WebView view, String url, Bitmap favicon) {
            parseUrl(url);
        }
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/activity/ValidateActivity.kt
package dev.ragnarok.fenrir.activity

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.ViewGroup
import android.webkit.CookieManager
import android.webkit.RenderProcessGoneDetail
import android.webkit.WebView
import androidx.activity.OnBackPressedCallback
import androidx.activity.SystemBarStyle
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.insets.ProtectionLayout
import androidx.core.view.iterator
import androidx.webkit.WebViewClientCompat
import dev.ragnarok.fenrir.Extra
import dev.ragnarok.fenrir.Includes
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.UserAgentTool
import dev.ragnarok.fenrir.api.Auth
import dev.ragnarok.fenrir.api.util.VKStringUtils
import dev.ragnarok.fenrir.api.validation.IValidateProvider
import dev.ragnarok.fenrir.nonNullNoEmpty
import dev.ragnarok.fenrir.settings.ISettings
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.settings.theme.ThemesController.currentStyle
import dev.ragnarok.fenrir.util.Logger
import dev.ragnarok.fenrir.util.coroutines.CompositeJob
import dev.ragnarok.fenrir.util.coroutines.CoroutinesUtils.sharedFlowToMain
import dev.ragnarok.fenrir.util.toast.CustomToast.Companion.createCustomToast
import kotlinx.coroutines.flow.filter
import kotlin.math.max

class ValidateActivity : AppCompatActivity() {
    private var validateProvider: IValidateProvider? = null
    private val mCompositeJob = CompositeJob()
    private var urlVal: String? = null
    private var accountId: Long = ISettings.IAccountsSettings.INVALID_ID

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(currentStyle())
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        urlVal = (intent.getStringExtra(EXTRA_VALIDATE) ?: return)
        accountId = intent.getLongExtra(Extra.ACCOUNT_ID, ISettings.IAccountsSettings.INVALID_ID)

        validateProvider = Includes.validationProvider

        validateProvider?.let {
            mCompositeJob.add(
                it.observeWaiting()
                    .filter { ob -> ob == urlVal }
                    .sharedFlowToMain { onWaitingRequestReceived() }
            )
            mCompositeJob.add(
                it.observeCanceling()
                    .filter { ob -> ob == urlVal }
                    .sharedFlowToMain { onRequestCancelled() }
            )
        }

        val webview = findViewById<WebView>(R.id.item_web_auth)
        webview.settings.javaScriptEnabled = true
        webview.settings.domStorageEnabled = true
        webview.settings.blockNetworkLoads = false
        webview.settings.blockNetworkImage = false
        webview.clearCache(true)
        webview.settings.userAgentString = UserAgentTool.getAccountUserAgent(accountId)

        //Чтобы получать уведомления об окончании загрузки страницы
        webview.webViewClient = object : WebViewClientCompat() {
            override fun onRenderProcessGone(
                view: WebView?,
                detail: RenderProcessGoneDetail?
            ): Boolean {
                webview.destroy()
                createCustomToast(
                    this@ValidateActivity,
                    null,
                    null
                )?.showToastError(R.string.crash_error_activity_out_of_memory)
                finish()
                return true
            }

            override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
                parseUrl(url)
            }
        }
        val cookieManager = CookieManager.getInstance()
        cookieManager.removeAllCookies {
            Log.d(
                TAG,
                "Cookie removed: $it"
            )
        }
        webview.loadUrl(urlVal ?: "")
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                cancel()
            }
        })

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.item_root)) { v, windowInsets ->
            val insets =
                windowInsets.getInsets(WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout())
            val imeFixedBottom =
                if (windowInsets.isVisible(WindowInsetsCompat.Type.ime())) max(
                    windowInsets.getInsets(
                        WindowInsetsCompat.Type.ime()
                    ).bottom, insets.bottom
                ) else insets.bottom
            v.setPadding(
                insets.left, insets.top,
                insets.right,
                imeFixedBottom
            )
            WindowInsetsCompat.CONSUMED
        }

        val statusBarColor = Color.TRANSPARENT
        val navigationBarColor = Color.TRANSPARENT
        val invertIcons = !Settings.get().ui().isDarkModeEnabled(
            this
        )
        val statusBarStyle = if (invertIcons) SystemBarStyle.light(
            statusBarColor,
            statusBarColor
        ) else SystemBarStyle.dark(statusBarColor)
        val navigationBarStyle = if (invertIcons) SystemBarStyle.light(
            navigationBarColor,
            navigationBarColor
        ) else SystemBarStyle.dark(navigationBarColor)
        for (i in (window.decorView as ViewGroup)) {
            if (i is ProtectionLayout) {
                (window.decorView as ViewGroup).removeView(i)
            }
        }
        enableEdgeToEdge(statusBarStyle, navigationBarStyle)
    }

    internal fun cancel() {
        urlVal?.let { validateProvider?.cancel(it) }
        supportFinishAfterTransition()
    }

    private fun onRequestCancelled() {
        supportFinishAfterTransition()
    }

    private fun onWaitingRequestReceived() {
        urlVal?.let { validateProvider?.notifyThatValidateEntryActive(it) }
    }

    override fun onDestroy() {
        mCompositeJob.cancel()
        super.onDestroy()
    }

    internal fun parseUrl(url: String?) {
        try {
            if (url == null) {
                return
            }
            Logger.d(TAG, "url=$url")
            if (url.startsWith(Auth.redirect_url)) {
                if (!url.contains("error=")) {
                    val intent = Intent()
                    try {
                        val accessToken = tryExtractAccessToken(url)
                        val userId = tryExtractUserId(url)
                        if (accessToken.nonNullNoEmpty() || userId.nonNullNoEmpty()) {
                            userId?.toLong()
                                ?.let {
                                    Settings.get().accounts().storeAccessToken(it, accessToken)
                                }
                        }
                    } catch (_: Exception) {
                    }
                    setResult(RESULT_OK, intent)
                }
                urlVal?.let { validateProvider?.enterState(it, true) }
                finish()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    companion object {
        private val TAG = ValidateActivity::class.simpleName.orEmpty()
        private const val EXTRA_VALIDATE = "validate"


        fun createIntent(context: Context, validate_url: String?, accountId: Long): Intent {
            return Intent(context, ValidateActivity::class.java)
                .putExtra(EXTRA_VALIDATE, validate_url)
                .putExtra(Extra.ACCOUNT_ID, accountId)
        }

        internal fun tryExtractAccessToken(url: String): String? {
            return VKStringUtils.extractPattern(url, "access_token=(.*?)&")
        }

        internal fun tryExtractUserId(url: String): String? {
            return VKStringUtils.extractPattern(url, "user_id=(\\d*)")
        }
    }
}